File: backend/app/routers/analisis.py
"""Endpoints de la investigación temática: hallazgos, tablas y capas de análisis.

Sirven los productos de ``data/processed/analisis/`` (generados por los scripts
de ``etl/analisis/``). El catálogo es dinámico: los archivos nuevos aparecen sin
reiniciar el API. Se leen en cada petición (son pequeños y locales).
"""
from __future__ import annotations

import json
import re
from pathlib import Path

import pandas as pd
from fastapi import APIRouter, HTTPException, Query

from ..config import DATA_DIR

router = APIRouter(prefix="/analisis", tags=["analisis"])

_ID_SEGURO = re.compile(r"^[a-z0-9_]+(/[a-z0-9_]+)?$")


def _dir_analisis() -> Path:
    carpeta = DATA_DIR / "analisis"
    if not carpeta.is_dir():
        raise HTTPException(404, "No hay resultados de análisis generados todavía.")
    return carpeta


def _resolver(dataset: str, extension: str) -> Path:
    """Convierte un id como 'cartografia/mineria_deforestacion' en ruta segura."""
    if not _ID_SEGURO.match(dataset):
        raise HTTPException(422, f"Identificador de dataset inválido: '{dataset}'.")
    ruta = (_dir_analisis() / dataset).with_suffix(extension)
    if not ruta.is_file():
        raise HTTPException(404, f"No existe el dataset '{dataset}{extension}'.")
    return ruta


@router.get("")
def catalogo() -> dict:
    """Catálogo de productos de análisis disponibles (tablas, resúmenes y capas)."""
    carpeta = _dir_analisis()
    tablas, resumenes, capas = [], [], []
    for ruta in sorted(carpeta.rglob("*")):
        if not ruta.is_file():
            continue
        rel = ruta.relative_to(carpeta).as_posix()
        id_ = rel.rsplit(".", 1)[0]
        if ruta.suffix == ".csv":
            tablas.append({"id": id_, "archivo": rel})
        elif ruta.suffix == ".geojson":
            capas.append({"id": id_, "archivo": rel})
        elif ruta.suffix == ".json" and ruta.stem != "hallazgos":
            resumenes.append({"id": id_, "archivo": rel})
    return {"tablas": tablas, "resumenes": resumenes, "capas": capas}


@router.get("/hallazgos")
def hallazgos() -> dict:
    """Hallazgos consolidados de todas las líneas de investigación.

    Une ``hallazgos.json`` (paquete de monitoreo) con
    ``hallazgos_cartografia.json`` (cartografía oficial), ordenados por
    relevancia descendente.
    """
    carpeta = _dir_analisis()
    todos: list[dict] = []
    for nombre in ("hallazgos.json", "hallazgos_cartografia.json"):
        ruta = carpeta / nombre
        if ruta.is_file():
            try:
                todos.extend(json.loads(ruta.read_text(encoding="utf-8-sig")))
            except (json.JSONDecodeError, UnicodeDecodeError) as exc:
                raise HTTPException(500, f"{nombre} corrupto: {exc}") from exc
    if not todos:
        raise HTTPException(404, "Aún no hay hallazgos publicados.")
    todos.sort(key=lambda h: h.get("relevancia", 0), reverse=True)
    return {"hallazgos": todos, "total": len(todos)}


@router.get("/tabla/{dataset:path}")
def tabla(
    dataset: str,
    municipio: str | None = Query(None, description="Filtra si la tabla tiene columna municipio"),
    periodo: str | None = Query(None, description="Filtra si la tabla tiene columna periodo"),
    limite: int = Query(5000, ge=1, le=50000),
) -> dict:
    """Tabla de análisis en JSON (filas), con filtros básicos opcionales."""
    ruta = _resolver(dataset, ".csv")
    df = pd.read_csv(ruta, encoding="utf-8-sig")
    if municipio and "municipio" in df.columns:
        objetivo = municipio.strip().casefold()
        df = df[df["municipio"].astype(str).str.casefold() == objetivo]
    if periodo and "periodo" in df.columns:
        df = df[df["periodo"].astype(str) == periodo]
    total = int(len(df))
    df = df.head(limite)
    df = df.where(pd.notna(df), None)
    return {
        "dataset": dataset,
        "columnas": list(df.columns),
        "filas": df.to_dict(orient="records"),
        "total_filas": total,
        "truncado": total > limite,
    }


@router.get("/resumen/{dataset:path}")
def resumen(dataset: str) -> dict:
    """Resumen JSON de una línea de análisis (agregados clave del minero)."""
    ruta = _resolver(dataset, ".json")
    return json.loads(ruta.read_text(encoding="utf-8-sig"))


@router.get("/geo/{dataset:path}")
def geo(dataset: str) -> dict:
    """Capa GeoJSON de análisis (p. ej. 'recurrencia' — frentes persistentes)."""
    ruta = _resolver(dataset, ".geojson")
    return json.loads(ruta.read_text(encoding="utf-8"))


# --- Deforestación por unidad territorial (áreas protegidas, POMCAS, etc.) ----

# Configuración por tipo: archivo CSV, columna de nombre, columna secundaria,
# y de dónde sale la deforestación (columna directa o filtro por clase).
_TERRITORIOS: dict[str, dict] = {
    "areas_protegidas": {
        "titulo": "Áreas protegidas",
        "archivo": "areas_protegidas_serie.csv",
        "col_nombre": "nombre",
        "col_sec": "categoria",
        "por_clase": True,
        "col_area": None,
    },
    "resguardos": {
        "titulo": "Resguardos indígenas",
        "archivo": "cartografia/territorios_oficiales.csv",
        "filtro": ("tipo", "resguardo"),
        "col_nombre": "nombre",
        "col_sec": "pueblo",
        "col_defo": "deforestacion_ha",
        "col_area": "area_oficial_ha",
    },
    "consejos": {
        "titulo": "Consejos comunitarios",
        "archivo": "cartografia/territorios_oficiales.csv",
        "filtro": ("tipo", "consejo_comunitario"),
        "col_nombre": "nombre",
        "col_sec": "municipios",
        "col_defo": "deforestacion_ha",
        "col_area": "area_oficial_ha",
    },
    "pomcas": {
        "titulo": "POMCAS (cuencas ordenadas)",
        "archivo": "cartografia/pomcas_serie.csv",
        "col_nombre": "pomca",
        "col_sec": None,
        "col_defo": "deforestacion_ha",
        "col_area": None,
    },
    "cuencas": {
        "titulo": "Cuencas hidrográficas",
        "archivo": "cuencas_serie.csv",
        "col_nombre": "cuenca",
        "col_sec": None,
        "por_clase": True,
        "col_area": None,
    },
}


@router.get("/territorios")
def territorios_catalogo() -> dict:
    """Tipos de unidad territorial disponibles y si tienen datos generados."""
    carpeta = _dir_analisis()
    tipos = []
    for tid, cfg in _TERRITORIOS.items():
        ruta = (carpeta / cfg["archivo"])
        tipos.append({"id": tid, "titulo": cfg["titulo"], "disponible": ruta.is_file()})
    return {"tipos": tipos}


@router.get("/territorios/{tipo}")
def territorios(
    tipo: str,
    periodo: str | None = Query(None, description="Filtra a un periodo; si se omite, acumulado"),
) -> dict:
    """Ranking de deforestación por unidad para un tipo de territorio.

    Devuelve cada unidad con su deforestación mapeada (acumulada o de un periodo
    concreto), número de periodos con dato, área oficial y % del territorio,
    ordenadas de mayor a menor. Incluye la lista de periodos disponibles para
    alimentar la barra de tiempo del dashboard.
    """
    cfg = _TERRITORIOS.get(tipo)
    if cfg is None:
        raise HTTPException(
            422, f"Tipo desconocido: '{tipo}'. Disponibles: {', '.join(_TERRITORIOS)}."
        )
    ruta = _dir_analisis() / cfg["archivo"]
    if not ruta.is_file():
        raise HTTPException(404, f"Aún no hay datos de '{tipo}' ({cfg['archivo']}).")
    df = pd.read_csv(ruta, encoding="utf-8-sig")

    if cfg.get("filtro"):
        col, val = cfg["filtro"]
        df = df[df[col].astype(str).str.strip().str.lower() == val.lower()]

    nombre = cfg["col_nombre"]
    if cfg.get("por_clase"):
        df = df[df["clase"] == "Deforestación"].copy()
        df["_defo"] = df["hectareas"]
    else:
        df["_defo"] = df[cfg["col_defo"]]

    df = df.dropna(subset=[nombre])

    # Periodos disponibles (antes de filtrar) para la barra de tiempo
    periodos_disponibles = (
        sorted(str(p) for p in df["periodo"].dropna().unique()) if "periodo" in df else []
    )
    if periodo and "periodo" in df.columns:
        df = df[df["periodo"].astype(str) == periodo]

    unidades = []
    for clave, grp in df.groupby(nombre):
        total = float(grp["_defo"].sum())
        sec = None
        if cfg.get("col_sec") and cfg["col_sec"] in grp.columns:
            valores = [str(v) for v in grp[cfg["col_sec"]].dropna().unique() if str(v) != "nan"]
            sec = valores[0] if valores else None
        area = None
        pct = None
        if cfg.get("col_area") and cfg["col_area"] in grp.columns:
            area_vals = grp[cfg["col_area"]].dropna()
            if len(area_vals):
                area = round(float(area_vals.iloc[0]), 1)
                if area:
                    pct = round(100 * total / area, 3)
        unidades.append({
            "nombre": str(clave),
            "detalle": sec,
            "deforestacion_ha": round(total, 2),
            "n_periodos": int(grp["periodo"].nunique()) if "periodo" in grp else None,
            "area_ha": area,
            "pct_del_territorio": pct,
        })
    unidades.sort(key=lambda x: x["deforestacion_ha"], reverse=True)
    return {
        "tipo": tipo,
        "titulo": cfg["titulo"],
        "periodo": periodo,
        "periodos_disponibles": periodos_disponibles,
        "n_unidades": len(unidades),
        "deforestacion_total_ha": round(sum(u["deforestacion_ha"] for u in unidades), 1),
        "nota": "Deforestación mapeada dentro de cada unidad. Ver metodología en /datos.",
        "unidades": unidades,
    }

File: frontend/src/server/analisis.ts
/**
 * Lógica de la investigación temática (réplica de routers/analisis.py):
 * catálogo, hallazgos, tablas, resúmenes, capas geo, parches y territorios.
 */
import fs from 'node:fs';
import path from 'node:path';
import { DATA_DIR, datos, ErrorDatos, normalizar, parseCsv, type FeatureCollection } from './datos';

const DIR_ANALISIS = path.join(DATA_DIR, 'analisis');
const ID_SEGURO = /^[a-z0-9_]+(\/[a-z0-9_]+)?$/;

function existeDirAnalisis(): string {
  if (!fs.existsSync(DIR_ANALISIS)) throw new ErrorDatos('No hay resultados de análisis generados todavía.', 404);
  return DIR_ANALISIS;
}

function resolver(dataset: string, ext: string): string {
  if (!ID_SEGURO.test(dataset)) throw new ErrorDatos(`Identificador de dataset inválido: '${dataset}'.`, 422);
  const ruta = path.join(existeDirAnalisis(), dataset + ext);
  if (!fs.existsSync(ruta)) throw new ErrorDatos(`No existe el dataset '${dataset}${ext}'.`, 404);
  return ruta;
}

export function catalogoAnalisis() {
  const carpeta = existeDirAnalisis();
  const tablas: { id: string; archivo: string }[] = [];
  const resumenes: { id: string; archivo: string }[] = [];
  const capas: { id: string; archivo: string }[] = [];
  const walk = (dir: string, base = ''): void => {
    for (const nombre of fs.readdirSync(dir).sort()) {
      const abs = path.join(dir, nombre);
      const rel = base ? `${base}/${nombre}` : nombre;
      if (fs.statSync(abs).isDirectory()) { walk(abs, rel); continue; }
      const id = rel.replace(/\.[^.]+$/, '');
      if (nombre.endsWith('.csv')) tablas.push({ id, archivo: rel });
      else if (nombre.endsWith('.geojson')) capas.push({ id, archivo: rel });
      else if (nombre.endsWith('.json') && !nombre.startsWith('hallazgos')) resumenes.push({ id, archivo: rel });
    }
  };
  walk(carpeta);
  return { tablas, resumenes, capas };
}

export function hallazgos() {
  const carpeta = existeDirAnalisis();
  const todos: Record<string, unknown>[] = [];
  for (const nombre of ['hallazgos.json', 'hallazgos_cartografia.json']) {
    const ruta = path.join(carpeta, nombre);
    if (fs.existsSync(ruta)) {
      todos.push(...(JSON.parse(fs.readFileSync(ruta, 'utf-8')) as Record<string, unknown>[]));
    }
  }
  if (!todos.length) throw new ErrorDatos('Aún no hay hallazgos publicados.', 404);
  todos.sort((a, b) => Number(b.relevancia ?? 0) - Number(a.relevancia ?? 0));
  return { hallazgos: todos, total: todos.length };
}

export function tablaAnalisis(dataset: string, opts: { municipio?: string | null; periodo?: string | null; limite?: number }) {
  const ruta = resolver(dataset, '.csv');
  let filas = parseCsv(fs.readFileSync(ruta, 'utf-8'));
  const columnas = filas.length ? Object.keys(filas[0]) : [];
  if (opts.municipio && columnas.includes('municipio')) {
    const obj = opts.municipio.trim().toLowerCase();
    filas = filas.filter((f) => String(f.municipio).toLowerCase() === obj);
  }
  if (opts.periodo && columnas.includes('periodo')) {
    filas = filas.filter((f) => String(f.periodo) === opts.periodo);
  }
  const total = filas.length;
  const limite = opts.limite ?? 5000;
  const recortadas = filas.slice(0, limite).map((f) => {
    const o: Record<string, unknown> = {};
    for (const c of columnas) {
      const v = f[c];
      const n = Number(v);
      o[c] = v === '' ? null : (v !== '' && Number.isFinite(n) && /^-?\d/.test(v) ? n : v);
    }
    return o;
  });
  return { dataset, columnas, filas: recortadas, total_filas: total, truncado: total > limite };
}

export function resumenAnalisis(dataset: string): unknown {
  return JSON.parse(fs.readFileSync(resolver(dataset, '.json'), 'utf-8'));
}
export function geoAnalisis(dataset: string): unknown {
  return JSON.parse(fs.readFileSync(resolver(dataset, '.geojson'), 'utf-8'));
}

// ── parches (réplica de repository.parches) ─────────────────────────────────
export function parches(periodo: string | null, minHa: number, municipio: string | null): FeatureCollection {
  const meta = (datos.metadata().periodos as { id: string; ano_inicio?: number }[]) ?? [];
  const anoPor = new Map(meta.map((p) => [p.id, p.ano_inicio]));
  let periodos: string[];
  if (periodo != null) {
    datos.validarPeriodo(periodo);
    periodos = datos.hotspotsDisponibles().includes(periodo) ? [periodo] : [];
  } else {
    periodos = datos.hotspotsDisponibles();
  }
  let objetivoMun: string | null = null;
  if (municipio) {
    try { objetivoMun = normalizar(datos.nombreMunicipio(datos.resolverMunicipio(municipio))); }
    catch { objetivoMun = normalizar(municipio); }
  }
  const features: unknown[] = [];
  let totalHa = 0;
  const porPeriodo = new Map<string, { periodo: string; ano_inicio: number | undefined; n: number; ha: number }>();
  for (const pid of periodos) {
    const fc = datos.hotspots(pid);
    if (!fc) continue;
    const ano = anoPor.get(pid);
    for (const feat of fc.features as { geometry?: unknown; properties?: Record<string, unknown> }[]) {
      const props = feat.properties ?? {};
      const ha = Number(props.ha) || 0;
      if (ha < minHa) continue;
      const mun = props.municipio ?? null;
      if (objetivoMun != null && (mun == null || normalizar(String(mun)) !== objetivoMun)) continue;
      features.push({ type: 'Feature', geometry: feat.geometry, properties: { periodo: pid, ano_inicio: ano, municipio: mun, ha: Math.round(ha * 100) / 100 } });
      totalHa += ha;
      const agg = porPeriodo.get(pid) ?? { periodo: pid, ano_inicio: ano, n: 0, ha: 0 };
      agg.n += 1; agg.ha += ha;
      porPeriodo.set(pid, agg);
    }
  }
  const porPeriodoArr = [...porPeriodo.values()].map((a) => ({ ...a, ha: Math.round(a.ha * 10) / 10 }))
    .sort((a, b) => (a.ano_inicio ?? 0) - (b.ano_inicio ?? 0));
  return {
    type: 'FeatureCollection', features,
    metadata: { n_parches: features.length, ha_total: Math.round(totalHa * 10) / 10, periodos, min_ha: minHa, por_periodo: porPeriodoArr },
  };
}

/**
 * Resumen ligero (sin geometría) de los polígonos de deforestación por periodo,
 * para dibujar la línea de tiempo del visor. Devuelve conteo y área de cada
 * periodo con hotspots, opcionalmente acotado a un municipio.
 */
export function parchesResumen(municipio: string | null): {
  por_periodo: { periodo: string; ano_inicio: number | undefined; n: number; ha: number }[];
} {
  const meta = (datos.metadata().periodos as { id: string; ano_inicio?: number }[]) ?? [];
  const anoPor = new Map(meta.map((p) => [p.id, p.ano_inicio]));
  let objetivoMun: string | null = null;
  if (municipio) {
    try { objetivoMun = normalizar(datos.nombreMunicipio(datos.resolverMunicipio(municipio))); }
    catch { objetivoMun = normalizar(municipio); }
  }
  const filas: { periodo: string; ano_inicio: number | undefined; n: number; ha: number }[] = [];
  for (const pid of datos.hotspotsDisponibles()) {
    const fc = datos.hotspots(pid);
    if (!fc) continue;
    let n = 0;
    let ha = 0;
    for (const feat of fc.features as { properties?: Record<string, unknown> }[]) {
      const props = feat.properties ?? {};
      if (objetivoMun != null) {
        const mun = props.municipio ?? null;
        if (mun == null || normalizar(String(mun)) !== objetivoMun) continue;
      }
      n += 1;
      ha += Number(props.ha) || 0;
    }
    filas.push({ periodo: pid, ano_inicio: anoPor.get(pid), n, ha: Math.round(ha * 10) / 10 });
  }
  filas.sort((a, b) => (a.ano_inicio ?? 0) - (b.ano_inicio ?? 0));
  return { por_periodo: filas };
}

// ── territorios (réplica de /analisis/territorios) ──────────────────────────
interface CfgTerritorio {
  titulo: string; archivo: string; col_nombre: string; col_sec?: string | null;
  filtro?: [string, string]; por_clase?: boolean; col_defo?: string; col_area?: string | null;
}
export const TERRITORIOS: Record<string, CfgTerritorio> = {
  areas_protegidas: { titulo: 'Áreas protegidas', archivo: 'areas_protegidas_serie.csv', col_nombre: 'nombre', col_sec: 'categoria', por_clase: true, col_area: null },
  resguardos: { titulo: 'Resguardos indígenas', archivo: 'cartografia/territorios_oficiales.csv', filtro: ['tipo', 'resguardo'], col_nombre: 'nombre', col_sec: 'pueblo', col_defo: 'deforestacion_ha', col_area: 'area_oficial_ha' },
  consejos: { titulo: 'Consejos comunitarios', archivo: 'cartografia/territorios_oficiales.csv', filtro: ['tipo', 'consejo_comunitario'], col_nombre: 'nombre', col_sec: 'municipios', col_defo: 'deforestacion_ha', col_area: 'area_oficial_ha' },
  pomcas: { titulo: 'POMCAS (cuencas ordenadas)', archivo: 'cartografia/pomcas_serie.csv', col_nombre: 'pomca', col_sec: null, col_defo: 'deforestacion_ha', col_area: null },
  cuencas: { titulo: 'Cuencas hidrográficas', archivo: 'cuencas_serie.csv', col_nombre: 'cuenca', col_sec: null, por_clase: true, col_area: null },
};

export function catalogoTerritorios() {
  const carpeta = existeDirAnalisis();
  return {
    tipos: Object.entries(TERRITORIOS).map(([id, cfg]) => ({
      id, titulo: cfg.titulo, disponible: fs.existsSync(path.join(carpeta, cfg.archivo)),
    })),
  };
}

export function territorios(tipo: string, periodo: string | null) {
  const cfg = TERRITORIOS[tipo];
  if (!cfg) throw new ErrorDatos(`Tipo desconocido: '${tipo}'. Disponibles: ${Object.keys(TERRITORIOS).join(', ')}.`, 422);
  const ruta = path.join(existeDirAnalisis(), cfg.archivo);
  if (!fs.existsSync(ruta)) throw new ErrorDatos(`Aún no hay datos de '${tipo}' (${cfg.archivo}).`, 404);
  let filas = parseCsv(fs.readFileSync(ruta, 'utf-8'));

  if (cfg.filtro) {
    const [col, val] = cfg.filtro;
    filas = filas.filter((f) => String(f[col]).trim().toLowerCase() === val.toLowerCase());
  }
  filas = filas.filter((f) => f[cfg.col_nombre] != null && f[cfg.col_nombre] !== '');
  if (cfg.por_clase) filas = filas.filter((f) => f.clase === 'Deforestación');

  const periodosDisp = [...new Set(filas.map((f) => f.periodo).filter(Boolean))].sort();
  if (periodo) filas = filas.filter((f) => String(f.periodo) === periodo);

  const grupos = new Map<string, Record<string, string>[]>();
  for (const f of filas) {
    const clave = String(f[cfg.col_nombre]);
    (grupos.get(clave) ?? grupos.set(clave, []).get(clave)!).push(f);
  }
  const defoCol = cfg.por_clase ? 'hectareas' : (cfg.col_defo as string);
  const unidades = [...grupos.entries()].map(([nombre, grp]) => {
    const total = grp.reduce((s, r) => s + (Number(r[defoCol]) || 0), 0);
    let detalle: string | null = null;
    if (cfg.col_sec && grp[0][cfg.col_sec] != null) {
      const vals = grp.map((r) => r[cfg.col_sec!]).filter((v) => v && v !== 'nan' && v !== '');
      detalle = vals.length ? String(vals[0]) : null;
    }
    let area: number | null = null;
    let pct: number | null = null;
    if (cfg.col_area && grp[0][cfg.col_area] != null && grp[0][cfg.col_area] !== '') {
      area = Math.round(Number(grp[0][cfg.col_area]) * 10) / 10;
      if (area) pct = Math.round((100 * total / area) * 1000) / 1000;
    }
    const nPeriodos = new Set(grp.map((r) => r.periodo).filter(Boolean)).size || null;
    return { nombre, detalle, deforestacion_ha: Math.round(total * 100) / 100, n_periodos: nPeriodos, area_ha: area, pct_del_territorio: pct };
  }).sort((a, b) => b.deforestacion_ha - a.deforestacion_ha);

  return {
    tipo, titulo: cfg.titulo, periodo, periodos_disponibles: periodosDisp,
    n_unidades: unidades.length,
    deforestacion_total_ha: Math.round(unidades.reduce((s, u) => s + u.deforestacion_ha, 0) * 10) / 10,
    nota: 'Deforestación mapeada dentro de cada unidad. Ver metodología en /datos.',
    unidades,
  };
}
